diff -uparN libmetal-2016.10/lib/config.h libmetal/lib/config.h
--- libmetal-2016.10/lib/config.h	2016-10-14 19:43:45.000000000 -0400
+++ libmetal/lib/config.h	2019-07-26 18:26:40.454139842 -0400
@@ -64,7 +64,7 @@ extern "C" {
 #define METAL_MACHINE		"@PROJECT_MACHINE@"
 #define METAL_MACHINE_@PROJECT_MACHINE_UPPER@
 
-#cmakedefine HAVE_STDATOMIC_H
+#define HAVE_STDATOMIC_H
 #cmakedefine HAVE_FUTEX_H
 
 #ifdef __cplusplus
